WatsonTcp/WatsonCommon.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WatsonTcp
+{
+    internal static class WatsonCommon
+    {
+        internal static byte[] ReadStreamFully(Stream input)
+        { 
+            byte[] buffer = new byte[65536];
+            using (MemoryStream ms = new MemoryStream())
+            {
+                int read = 0;
+                while (true)
+                {
+                    read = input.Read(buffer, 0, buffer.Length);
+                    if (read > 0)
+                    {
+                        ms.Write(buffer, 0, read);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+
+                return ms.ToArray();
+            }
+        }
+
+        internal static byte[] ReadFromStream(Stream stream, long count, int bufferLen)
+        {
+            if (count <= 0) return null;
+            if (bufferLen <= 0) throw new ArgumentException("Buffer must be greater than zero bytes."); 
+            byte[] buffer = new byte[bufferLen];
+
+            int read = 0;
+            long bytesRemaining = count;
+            MemoryStream ms = new MemoryStream();
+
+            while (bytesRemaining > 0)
+            {
+                if (bufferLen > bytesRemaining) buffer = new byte[bytesRemaining];
+
+                read = stream.Read(buffer, 0, buffer.Length);
+                if (read > 0)
+                {
+                    ms.Write(buffer, 0, read);
+                    bytesRemaining -= read;
+                }
+                else
+                {
+                    throw new IOException("Could not read from supplied stream.");
+                }
+            }
+
+            byte[] data = ms.ToArray();
+            return data;
+        }
+
+        internal static async Task<byte[]> ReadFromStreamAsync(Stream stream, long count, int bufferLen)
+        {
+            if (count <= 0) return null;
+            if (bufferLen <= 0) throw new ArgumentException("Buffer must be greater than zero bytes.");
+            byte[] buffer = new byte[bufferLen];
+
+            int read = 0;
+            long bytesRemaining = count;
+            MemoryStream ms = new MemoryStream();
```

Comment:
Here is another memory leak

Reply:
This is intentional. The idea is to allow the consumer to close and dispose it.

Reply:
I fixed the leak in WatsonCommon.cs with NuGet v4.1.12.3: https://www.nuget.org/packages/WatsonTcp/4.1.12.3

The other one is in the PR so I'll leave that to @developervariety

Reply:
> I fixed the leak in WatsonCommon.cs with NuGet v4.1.12.3: https://www.nuget.org/packages/WatsonTcp/4.1.12.3
> 
> The other one is in the PR so I'll leave that to @developervariety

Thats a quick fix :+1: well done